text/0000-adaptive-hashing.md

```diff
@@ -0,0 +1,193 @@
+  // ...
+```
+
+Here’s a state diagram for HashMap with adaptive hashing. The dashed edge means the state change is
+very unlikely, and the dotted edge means the state change is enormously unlikely.
+
+<img width="800" src="https://cdn.rawgit.com/pczarn/code/d62cd067ca84ff049ef196aa1b7773d67b4189d4/rust/robinhood/adaptive.svg">
+
+## Choosing constants
+
+The thresholds of 128 and 512 are chosen to minimize the chance of exceeding them. In particular, we
+want that chance to be less than 10^-8 with a load of 90% and less than 10^-30 with a load of 20%.
+For displacement, the smallest k that fits our needs is 90, so we round that up to 128. For the
+number of forward-shifted buckets, we choose k=512. Keep in mind that the run length is a sum of the
+displacement and the number of forward-shifted buckets, so its threshold is 128+512=640. Even though
+the probability of having a run length of more than 640 buckets may be higher than the probability
+we want, it should be low enough.
+
+<img width="600" src="https://cdn.rawgit.com/pczarn/code/d62cd067ca84ff049ef196aa1b7773d67b4189d4/rust/robinhood/lookup_cost.png">
+
+<img width="600" src="https://cdn.rawgit.com/pczarn/code/d62cd067ca84ff049ef196aa1b7773d67b4189d4/rust/robinhood/run_length.png">
+
+## Choosing hash functions
+
+For hashing integers, the best choice is a mixer similar to the one used in SipHash’s finalizer. For
+strings and slices of integers, we will use FarmHash. (The Hasher trait must allow one-shot hashing
+for FarmHash.) Using any other key type means your HashMap will do safe hashing.
+
+# Consequences
+## For the performance of Rust programs
+
+The impact is minimal on programs that rarely use HashMaps. The increase in binary size should be
+small. For programs that spend a large portion of their run time using HashMap with primitive keys,
+the speedup should be noticeable.
+
+On 32-bit platforms, the benefit of using a 32-bit hash function instead of SipHash is higher,
+because SipHash’s round involves 30 32-bit operations.
+
+## For the HashMap API
+
+One day, we may want to hash HashMaps. The hashing infrastructure can be changed to allow it. The
+implementation of Hash for HashMap can hash the hashes stored in every map, rather than  However,
+adaptive hashing makes it harder to write a correct and performant implementation of Hash for
+HashMap. If two HashMaps (that can be compared) have equal values, they must hash to the same
+integer. However, with adaptive hashing, HashMap can switch to the safe mode, which means it no
+longer stores the same hashes as other HashMaps that remain in ‘fast’ mode. The only way to handle
+the situation for the safe mode is to rehash all keys as if the HashMap were in ‘fast’ mode, which
+may take a significant time.
+
+## For the order of iteration
+
+Currently, HashMaps have nondeterministic order of iteration by default. This is seen as a good
+thing, because programmers that test programs won’t learn to rely on a fixed iteration order.
```

Comment:
nit: “programmers that test programs” makes little grammatical sense. Maybe:

>  because programmers then ensure programs do not rely on a fixed iteration order